pallets/operations/src/impls.rs

```diff
@@ -1,67 +1,85 @@
+            expected_consumers += 1;
+            if has_staking_lock {
+                expected_consumers += 1;
+            }
+        }
+        if Self::has_next_session_keys_and_account_is_controller(&who) {
+            expected_consumers += 1;
         }
 
-        weight += T::DbWeight::get().reads(2);
-        if Self::staker_has_consumers_underflow(&who, consumers) {
+        if current_consumers < expected_consumers {
+            log::debug!(
+                target: LOG_TARGET,
+                "Account {:?} has current consumers {} less than expected consumers {:?}, incrementing ",
+                HexDisplay::from(&who.encode()), current_consumers, expected_consumers);
             Self::increment_consumers(who)?;
-            weight += T::DbWeight::get().writes(1);
-            return Ok(Some(weight).into());
+        } else {
+            log::debug!(
+                target: LOG_TARGET,
+                "Account {:?} does not have consumers underflow, not incrementing",
+                HexDisplay::from(&who.encode())
+            );
         }
 
-        log::debug!(
-            target: LOG_TARGET,
-            "Account {:?} has correct consumer counter, not incrementing",
-            HexDisplay::from(&who.encode())
-        );
-        Ok(Some(weight).into())
+        Ok(())
+    }
+
+    fn reserved_or_frozen_non_zero(who: &T::AccountId) -> bool {
+        !T::BalancesProvider::is_reserved_zero(who) || !T::BalancesProvider::is_frozen_zero(who)
     }
 
-    fn staker_has_consumers_underflow(who: &T::AccountId, consumers: u32) -> bool {
+    fn has_vesting_lock(who: &T::AccountId) -> bool {
         let locks = T::BalancesProvider::locks(who);
-        let has_vesting_lock = Self::has_lock(&locks, VESTING_ID);
-        let vester_has_consumers_underflow = consumers == 1 && has_vesting_lock;
-        let has_staking_lock = Self::has_lock(&locks, STAKING_ID);
-        let nominator_has_consumers_underflow = consumers == 2 && has_staking_lock;
-        let has_next_session_keys = T::NextKeysSessionProvider::has_next_session_keys(who);
-        let validator_has_consumers_underflow =
-            consumers == 3 && has_staking_lock && has_next_session_keys;
-        vester_has_consumers_underflow
-            || nominator_has_consumers_underflow
-            || validator_has_consumers_underflow
+        Self::has_lock(&locks, VESTING_ID)
     }
 
-    fn no_consumers_some_reserved(who: &T::AccountId, consumers: u32) -> bool {
-        let is_reserved_not_zero = T::BalancesProvider::is_reserved_not_zero(who);
+    fn has_staking_lock(who: &T::AccountId) -> bool {
+        let locks = T::BalancesProvider::locks(who);
+        Self::has_lock(&locks, STAKING_ID)
+    }
 
-        consumers == 0 && is_reserved_not_zero
+    fn has_next_session_keys_and_account_is_controller(who: &T::AccountId) -> bool {
+        let has_next_session_keys = T::NextKeysSessionProvider::has_next_session_keys(who);
```

Comment:
Why do we need to check if account is controller, instead of only if it has `session_keys`?

Reply:
From version 12 onwards, we would not have to check, yet in version 11, where it was still possible to set controller != stash, `session.set_keys`:
* retrieved **controller** account id via [this line](https://github.com/Cardinal-Cryptography/substrate/blob/aleph-v0.9.38/frame/session/src/lib.rs#L804), using [`ValidatorIdOf` trait](https://github.com/Cardinal-Cryptography/aleph-node/blob/r-11.4/bin/runtime/src/lib.rs#L392), that is a [`pallet_staking::StashOf` trait](https://github.com/Cardinal-Cryptography/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/staking/src/lib.rs#L924), which queries `staking.ledger` storage map. This means that input argument to `session.set_keys` is a controller account, which implies that [`consumers` counter is increased on the controller account](https://github.com/Cardinal-Cryptography/substrate/blob/aleph-v0.9.38/frame/session/src/lib.rs#L810), even though keys are stash always on a stash account.